uniffi_macros/src/export/trait_interface.rs

```diff
@@ -126,18 +127,37 @@ pub(crate) fn ffi_converter(
     let lift_ref_impl_spec = tagged_impl_header("LiftRef", &quote! { dyn #trait_ident }, remote);
     let trait_name = ident_to_string(trait_ident);
     let try_lift = if with_foreign {
+        // The trait can be implemented by both Rust and the foreign side.
+        // `try_lift` needs to handle both cases
         let trait_impl_ident = callback_interface::trait_impl_ident(&trait_name);
         quote! {
             fn try_lift(handle: Self::FfiType) -> ::uniffi::deps::anyhow::Result<::std::sync::Arc<Self>> {
-                ::std::result::Result::Ok(::std::sync::Arc::new(<#trait_impl_ident>::new(handle.as_raw())))
+                if handle.is_foreign() {
+                    // Handle was generated by the foreign side.
+                    // Construct a new instance of our struct that implements the trait using the handle and the registered vtable.
+                    ::std::result::Result::Ok(::std::sync::Arc::new(<#trait_impl_ident>::new(handle.as_raw())))
+                } else {
+                    // Handle was generated by the Rust side.
+                    // Reconstruct the double-wrapped arc that we lowered.
+                    use ::std::clone::Clone;
+                    // Note: handle is for a double-wrapped arc
+                    // https://mozilla.github.io/uniffi-rs/latest/internals/object_references.html
+                    let obj: ::std::sync::Arc<::std::sync::Arc<dyn #trait_ident>> = unsafe {
+                        handle.into_arc()
+                    };
+                    // We have a `Arc<Arc<dyn #trait_ident>>>`, but we need to return an `Arc<dyn #trait_ident>`
+                    // Unfortunately, Rust won't cast the double-arc to a single-arc trait object.
+                    // So we have to clone the inner arc and return that.
+                    ::std::result::Result::Ok((*obj).clone())
```

Comment:
This feels pretty weird and I have a couple ideas on how to fix it:

* Switch to using "wide handles" for trait interfaces.  This would be a struct with 2 `u64` handles.  This way we can represent the trait interface directly without having to create the extra arc.
* Implement something like `https://github.com/mozilla/uniffi-rs/pull/1808` where we manage the handles ourselves. 

I think I prefer the latter, but it's not clear if we want to go that route and I didn't want to block the trait interface fixes on that decision, so this PR only focuses on the trait interface wrapping issue.